### ISSUE TITLE
Unsupported Grant Type response on requesting with "client credentials"

### ISSUE THREAD

**Issue by maarten-shop4** [trikoder/oauth2-bundle#100]
I'm trying to connect to the `/authorize`  route in my (Symfony 4) application

This is the request data I'm sending:
```
{
    "grant_type": "client_credentials",
    "client_id": "test",
    "client_secret": "*********",
    "scope": "read"
}
```
with Headers: Content-Type = application/x-www-form-urlencoded

The response I'm getting back is:
```
{
    "error": "unsupported_grant_type",
    "error_description": "The authorization grant type is not supported by the authorization server.",
    "hint": "Check that all required parameters have been provided",
    "message": "The authorization grant type is not supported by the authorization server."
}
```
with Response code 400 "bad request"

I've added my client with this create-client command `php bin/console trikoder:o:create-client test ********** --scope=read --grant-type=client_credentials`

running `php bin/console trikoder:o:list-clients` shows this list:

| identifier | secret     | scope | redirect_uri | grant_type         |
|------------|------------|-------|--------------|--------------------|
| test       | ********** | read  |              | client_credentials |

What am I doing wrong, I have completely followed the installation guide.

**Comment by spideyfusion** [comment]
@maarten-shop4 Did you mean to send a request to the `/token` route?

The `/authorize` route is meant to be used with `authorization_code` and `implicit` grant types.

**Comment by maarten-shop4** [comment]
You're right, found this article about [oauth2 client credentials](https://aaronparecki.com/oauth-2-simplified/#client-credentials)

Unfortunately making a request to the `/token` route also doesn't give me the result I was hoping for.
I'd expect to receive some sort of token. All that I'm getting back is another `unsupported_grant_type` error.

My `POST` request to the `/token` route:
```
{
    grant_type: client_credentials,
    client_id: test,
    client_secret: *********
}
```

the response returned:
```JSON
{
     "error":"unsupported_grant_type",
     "error_description":"The authorization grant type is not supported by the authorization server.",
     "hint":"Check that all required parameters have been provided",
     "message":"The authorization grant type is not supported by the authorization server."
}
```

This is also weird because this is enabled in my config file `trikoder_oauth2.yaml`:
```YAML
trikoder_oauth2:
    authorization_server:
        # Full path to the private key file.
        # How to generate a private key: https://oauth2.thephpleague.com/installation/#generating-public-and-private-keys
        private_key: '%kernel.project_dir%/var/oauth/private.key'

        # Passphrase of the private key, if any
        private_key_passphrase: *****************

        # The plain string or the ascii safe string used to create a Defuse\Crypto\Key to be used as an encryption key.
        # How to generate an encryption key: https://oauth2.thephpleague.com/installation/#string-password
        encryption_key: '`*****************************'

        # The type of value of 'encryption_key'
        encryption_key_type: plain # One of "plain"; "defuse"

        # Token TTL values should be a valid interval: http://php.net/manual/en/dateinterval.construct.php#refsect1-dateinterval.construct-parameters

        # How long the issued access token should be valid for.
        access_token_ttl: PT12H
        # How long the issued refresh token should be valid for.
        refresh_token_ttl: P1M
        # How long the issued auth code should be valid for.
        auth_code_ttl: PT5M

        # Whether to enable the client credentials grant
        enable_client_credentials_grant: true

        # Whether to enable the password grant
        enable_password_grant: false

        # Whether to enable the refresh token grant
        enable_refresh_token_grant: false

        # Whether to enable the authorization code grant
        enable_auth_code_grant: true

        # Whether to enable the implicit grant
        enable_implicit_grant: false

    resource_server:
        # Full path to the public key file
        # How to generate a public key: https://oauth2.thephpleague.com/installation/#generating-public-and-private-keys
        public_key: '%kernel.project_dir%/var/oauth/public.key'

    # Scopes that you wish to utilize in your application.
    # This should be a simple array of strings.
    scopes: ['read', 'create', 'update', 'delete']

    # Configures different persistence methods that can be used by the bundle for saving client and token data.
    # Only one persistence method can be configured at a time.
    persistence:
        doctrine:
            entity_manager: default # Name of the entity manager that you wish to use for managing clients and tokens.
        #in_memory: true

    # The priority of the event listener that converts an Exception to a Response
    exception_event_listener_priority: 10
```

**Comment by spideyfusion** [comment]
@maarten-shop4 Can you send me the raw request you're trying to send? A cURL request like this should work for you:

```sh
curl https://example.url/token \
    --request POST \
    --data-urlencode 'client_id=test' \
    --data-urlencode 'client_secret=secret' \
    --data-urlencode 'grant_type=client_credentials'
```

Make sure you're sending the data as `application/x-www-form-urlencoded`.

**Comment by maarten-shop4** [comment]
thanks for the response. this helped me figure out what the problem was.
in stead of using postman I started using curl from the command line. `curl -X POST -H "Content-Type: application/x-www-form-urlencoded" -d "client_id=test&client_secret=**********&grant_type=client_credentials" my.domain.com/api/oauth/token`

The server was constantly telling me it couldn't handle a `GET` request, so I didn't understand how my `POST` request. until I noticed you were instructing curl to use the HTTPS protocol. this immediately solved the problem, sending a `POST` request to the server redirected the insecure request to `GET` request on the same path.

thanks for pointing my in the right direction. now let's see if I can fix the next error: `Not configuring explicitly the provider for the "oauth2" listener on "api" firewall is ambiguous as there is more than one registered provider.`
